### ISSUE TITLE
[xas_viewer] read multi-scans Spec files

### ISSUE THREAD

**Issue by maurov** [xraypy/xraylarch#283]
This is a task of #281.

@maurov Yeah, I'm happy to merge this.  

Do you have a couple of not-too-huge multi-spectrum spec files with XASF scans that you could share.  I was trying to figure this all out too.  I think it would be very nice if the "select columns for constructing mu(E)" data could work with a spec file, selecting which scan number, which array columns, etc.    And it looks like the silx.io code and your  DataSourceSpecH5 would basically make that easy -- mostly just down to wiring up the GUI code.

_Originally posted by @newville in https://github.com/xraypy/xraylarch/issues/282#issuecomment-769437638_

@newville An example of multi-spectrum Spec file is [here](https://github.com/maurov/xraysloth/blob/master/examples/specfiledata_tests.dat). You can use scans 71 (EXAFS) or 72-74 (XANES). Energy counter is `arr_hdh_ene`, detector is `zap_det_dtc` and the incoming beam normalization counter is `arr_I02sum`.

This is mainly a GUI-related task. I see two approaches here:

1) via `wxlib.columnframe` (simpler)

An easy way to catch if the file is a multi-scan Spec file is to check if the first line starts with `#L`. In this case the GUI can retrieve the list of scans to put in a combo box to show the `[scan number, title and date]` with a simple `DataSourceSpecH5(filename).get_scans()`, once the user selects the wanted scan, the scan number is feed  to `read_specfile` as it is for a single scan case.

2) via a clone of `wxlib.athena_importer (more elegant and reusable for future HDF5 files reading)

(I still need to look how to perform this)

@newville feel free to comment here which is your preferred way to proceed. If you have time to implement this soon, it would be great (four pushing our users to adopt `xas_viewer`), otherwise I will try to work on it next week.

**Comment by newville** [comment]
@maurov Thanks for the link to the spec file.  

> 2. via a clone of `wxlib.athena_importer (more elegant and reusable for future HDF5 files reading)

Ah, right.  Yes, I think you are right that a "spec importer" is probably the better approach - that would be a little more complicated than "athena importer" and more like "columnframe" but with ability to select scan by number/label and also select (and maybe rename and add together?) columns.   

Is it correct that if we use `silx.io` and `DataSourceSpecH5` that this same code could read the ESRF HDF5 files directly?   

Anyway, I'm happy to work on that GUI part of it, but I might go kind of slow ;).

**Comment by newville** [comment]
@maurov OK, I think there is a working imported for multiple scans for XAS_Viewer.   This is sort of a cross between the ASCII file reader ("select which arrays to use to build mu(E)") and the Athena Importer ("select which scans to import").  Of course, that assumes that all the imported scans use the same array/column layout.  I think that should be the most common situation, but I'm not sure what to do if there is some random "ascan" of some slit or something else in the middle -- the user sort of has to know which scans are actually useful data.

I'm going to tag this 0.9.51 push this to PyPI and other installers in the next couple of days.  I guess we can think of the Spec reading as "beta" for this release (it works for me with your one test Spec file), but let me know if works okay for you or if you have any suggestions.

**Comment by maurov** [comment]
@newville the single (actually, "first") scan Spec reading works at this stage and we can tag `0.9.51` as it is.

A truly multi-scan reading that would let the user import a bunch of scans in one action is not easy because, as you said, there may be "not useful" data in between the list of scans or the array/column layout may change. I think this could be left to a generic HDF5 importer in xas_viewer, because the Spec file mapping in `silx.io` is equivalent to a specific HDF5 mapping.

What I propose to implement here is to extend the ASCII file reader as "first select which scan to use and then which arrays to use to build mu(E)". This would go into `wxlib.columnframe`:

- if the file starts with `#L` (instead of `#S`):  generate a combo box with the list of scans where each entry shows `{scan_number}: {scan_title} ({scan_date})`. You get this list from `DataSourceSpecH5(filename).get_scans()`. By default, the first scan is selected (as it is now without the combo box).

![image](https://user-images.githubusercontent.com/545294/115504080-a6e89380-a277-11eb-9c14-4f5522a3b946.png)

- once the user selects the wanted scan, the scan number is feed to `read_specfile` as it is for the "first" scan case.

Do you think you could add this feature soon?

**Comment by newville** [comment]
@maurov Thanks!  Putting the scan title in the "Scan List" is a great idea - will do.

For testing whether a file was a Specfile, I was doing (see https://github.com/xraypy/xraylarch/blob/master/larch/io/specfile_reader.py#L273)
    if the file starts with "#F " or is HDF5, try to open it with `DataSourceSpecH5` and see if `get_scans()` succeeds and returns something other than `None`.    

If I change that "#F " or "#S " it reads the ABCF002.dat file.  That's easy enough.

I think this should also work with HDF5 "specfiles", but have not tried that.  It might be nice to know if the spec HDF5 files also have some attribute set or some other way to test for "spec"-ness.   Like, I have lots of different HDF5 formatted files (at least 3 formats that I use regularly) and none of them would be usable for this.  

So a couple of questions/comments might be: 
   a) does it work with HDF5 files?
   b) it would be nice to have a "convert all energy axes to eV", maybe happening in the data importer.
   b) should there be some sort of filtering of scans by title/scan type "show me only 'fscan*' scans", etc?  I'm sure that would vary a bit by beamline.

I think that b) and c) are doable but might wait a bit to find out what else might be needed.

**Comment by maurov** [comment]
> @maurov Thanks! Putting the scan title in the "Scan List" is a great idea - will do.

Great, thanks!
 
> For testing whether a file was a Specfile, I was doing (see https://github.com/xraypy/xraylarch/blob/master/larch/io/specfile_reader.py#L273)
> if the file starts with "#F " or is HDF5, try to open it with `DataSourceSpecH5` and see if `get_scans()` succeeds and returns something other than `None`.

Checking the "#F" start is the correct way to detect if it is a Spec file.

> If I change that "#F " or "#S " it reads the ABCF002.dat file. That's easy enough.

Please, do not change with "#S" in `is_specfile` because the `ABCF002.dat` is not a correctly formatted Spec file, but only a "cut" of a scan from a Spec file without the header. Those are more or less multi-columns ASCII files that may be read via `read_specfile` in the `columframe` importer.

> I think this should also work with HDF5 "specfiles", but have not tried that. It might be nice to know if the spec HDF5 files also have some attribute set or some other way to test for "spec"-ness. Like, I have lots of different HDF5 formatted files (at least 3 formats that I use regularly) and none of them would be usable for this.
> 
> So a couple of questions/comments might be:
> a) does it work with HDF5 files?

`DataSourceSpecH5` is a wrapper on top of `silx.io.open` that in case of Spec files uses the underlying `silx.io.spech5` to [map a Spec file as HDF5](http://www.silx.org/doc/silx/latest/modules/io/spech5.html). I use this wrapper to get easily the data arrays from a Spec file. In the case of a generic HDF5 the wrapper will not work because the data are stored in a different URL in the HDF5 file. Furthermore when `silx.io.open` reads an HDF5 file it simply return the `h5py.File` object, so `DataSourceSpecH5` will not work.

In conclusion, we may need a further iteration (or live chat) to decide how to build a generic HDF5 importer for xas_viewer.

> b) it would be nice to have a "convert all energy axes to eV", maybe happening in the data importer.

Currently the conversion happens when the data are imported as Groups in xas_viewer. I think it is fine for now.

> c) should there be some sort of filtering of scans by title/scan type "show me only 'fscan*' scans", etc? I'm sure that would vary a bit by beamline.

This is beamline dependent and we can live without filtering for now.

**Comment by maurov** [comment]
Everything seems to work so far. I have updated related improvements targeting next release.
